completion/zsh/_balena

```diff
@@ -0,0 +1,81 @@
+#compdef balena
+#autoload
+
+_balena() {
+  typeset -A opt_args
+  local context state line curcontext="$curcontext"
+
+ # Valid top-level completions
+  local commands; commands=(
+            app apps build config deploy device devices env envs help key \
+            keys local login logout logs note os preload quickstart settings \
+            scan ssh util version whoami
+  )
+
+  # Sub-completions
+  app_cmds=( create restart rm )
+  config_cmds=( generate inject read reconfigure write )
+  device_cmds=( identify init move public-url reboot register rename rm \
```

Comment:
The contents of file are mainly from the previous `balena-completion.bash`, right? Totally not your fault, but that file was quite outdated, missing new commands. Developers forget to update it. As it happens, now that we are fully using oclif, when `npm run build` is executed the following large-ish JSON file is automatically generated: `oclif.manifest.json` (It's not committed to the git repo, it's listed in `.gitignore`, but it's included in the CLI installers so you should be able to find it in the folder where the CLI is installed - except for the zip package...)
```
$ jq . < oclif.manifest.json
{
  "version": "12.40.0",
  "commands": {
    "apps": {
      "id": "apps",
      "description": "List all applications.\n\nlist all your balena applications.\n\nFor detailed information on a particular application,\nuse `balena app <application>` instead.",
      "usage": "apps",
      "pluginName": "balena-cli",
      "pluginType": "core",
      "aliases": [],
      "examples": [
        "$ balena apps"
      ],
      "flags": {
        "help": {
          "name": "help",
          "type": "boolean",
          "char": "h",
          "description": "show CLI help",
          "allowNo": false
        },
        "verbose": {
          "name": "verbose",
          "type": "boolean",
          "char": "v",
          "description": "No-op since release v12.0.0",
          "allowNo": false
        }
      },
      "args": []
    },
    "build": {
      "id": "build",
      "description": "Build a project locally.\n\n ...
      ...
```
The great thing about this JSON file is that it is auto generated in every CLI release (every time we run `npm run build`), and thus always up to date with all commands, subcommands and flags. I would like to tempt you to come up with a solution that auto generates this `_balena` file, having `oclif.manifest.json` as input. Do you know a bit of Javascript? The solution would be preferably written in Typescript or plain Javascript (Node.js). Say, this file:

`./completion/generate-completion.js`
* Input: `oclif.manifest.json`
* Output: `_balena`

Alternatively, `_balena` could dynamically load `oclif.manifest.json`, or some other auto generated file, but in this case, an extract of what matters should probably be cached somehow and not reloaded every time one hits the tab character on the command line, because `oclif.manifest.json` is quite chunky - 124KB in compact form.

Writing the solution in shell script may not be a good idea because of Windows compatibility: We would have to hook the code to run when `npm run build` is executed, which on Windows currently happens on `cmd.exe` (e.g. in balena CI prior to creation of the CLI's GUI installer for Windows). Ironically, we also want to have bash or zsh autocompletion on Windows, because end users may be using MSYS, Cygwin, WSL etc -- but that's after the CLI was built and installed.

Talking of Windows, files starting with an underscore, like `_balena`, are traditionally hidden files on Windows -- like files starting with a dot on Linux or macOS. For this reason, I suggest renaming `_balena` to something else. How about `completion.common` or `completion-common`? (Because it is used in common by `completion.bash` and `completion.zsh`.)

Reply:
Yep, I just copied the existing completion, adding a full regex from cmdline help completion file isn't a great first commit ^^

If you got a json that explain all cmdline options, I could forge the completion with it and add it as part of the install and update process. I will look at that.

Reply:
As a coincidence, while working on another PR (#2159), I came across the following `jq` expression that could be used for inspiration:
```bash
# print a list of top-level balena CLI commands
$ jq -r '.commands | keys | map(.[0:index(":")]) | unique | join(" ")' < oclif.manifest.json

api-key app apps build config deploy device devices env envs internal join key keys leave local login logout logs note orgs os preload push scan settings ssh support tag tags tunnel util version whoami
```
`jq` docs: https://stedolan.github.io/jq/manual/